ipa-core/src/protocol/ipa_prf/mod.rs

```diff
@@ -84,18 +89,36 @@ where
     F: PrimeField + ExtendableField,
     Replicated<F>: Serializable,
 {
-    // TODO (richaj): Add shuffle either before the protocol starts or, after converting match keys to elliptical curve.
-    // We might want to do it earlier as that's a cleaner code
+    let shuffled = shuffle_inputs(ctx.narrow(&Step::Shuffle), input_rows).await?;
+    let mut prfd_inputs =
+        compute_prf_for_inputs(ctx.narrow(&Step::ConvertInputRowsToPrf), shuffled).await?;
+
+    prfd_inputs.sort_by(|a, b| a.prf_of_match_key.cmp(&b.prf_of_match_key));
+    // Chunk the incoming stream of records into stream of vectors of records with the same PRF
+    let mut input_stream = stream_iter(prfd_inputs);
+    let first_row = input_stream.next().await;
+    if first_row.is_none() {
+        return Ok(vec![]);
+    }
+    let first_row = first_row.unwrap();
+    let rows_chunked_by_user = chunk_rows_by_user(input_stream, first_row);
```

Comment:
There is a bunch of identical looking code within `attribute_cap_aggregate`. Maybe you can just move the call to quicksort to be the first thing that happens within `evaluate_per_user_attribution_circuit`

Reply:
I did this but fighting with set_total_records here. How should that be set?